searchlib/src/vespa/searchlib/attribute/postinglistsearchcontext.hpp

```diff
@@ -67,12 +67,54 @@ PostingListSearchContextT<DataT>::fillArray()
     _merger.merge();
 }
 
+template <typename DataT>
+struct PostingListSearchContextT<DataT>::FillPart : public vespalib::Runnable {
+    FillPart(const PostingStore& posting_store, const DictionaryConstIterator & from, size_t count,
+             BitVector & bv, uint32_t limit)
+        : _posting_store(posting_store), _bv(bv),
+          _docIdLimit(limit),
+          _from(from),
+          _to(from)
+    {
+        _to += count;
+    }
+    void run() override {
+        for (;_from != _to;++_from) {
+            addToBitVector(PostingListTraverser<PostingStore>(_posting_store, _from.getData().load_acquire()));
+        }
+    }
+    void addToBitVector(const PostingListTraverser<PostingStore> & postingList)
+    {
+        postingList.foreach_key([this](uint32_t key) {
+            if (__builtin_expect(key < _docIdLimit, true)) { _bv.setBit(key); }
+        });
+    }
+    const PostingStore      &_posting_store;
+    BitVector               &_bv;
+    uint32_t                 _docIdLimit;
+    DictionaryConstIterator  _from;
+    DictionaryConstIterator  _to;
+};
+
 template <typename DataT>
 void
-PostingListSearchContextT<DataT>::fillBitVector()
+PostingListSearchContextT<DataT>::fillBitVector(vespalib::ThreadBundle & thread_bundle)
 {
-    for (auto it(_lowerDictItr); it != _upperDictItr; ++it) {
-        _merger.addToBitVector(PostingListTraverser<PostingStore>(_posting_store, it.getData().load_acquire()));
+    size_t num_iter = _upperDictItr - _lowerDictItr;
+    size_t num_per_thread = (num_iter + (thread_bundle.size() - 1))/ thread_bundle.size();
+    std::vector<FillPart> parts;
+    parts.reserve(thread_bundle.size());
+    BitVector & master = *_merger.getBitVector();
+    std::vector<std::unique_ptr<BitVector>> scratch_bvs;
+    parts.emplace_back(_posting_store, _lowerDictItr, num_per_thread, master, _merger.getDocIdLimit());
+    for (size_t i(1); i < thread_bundle.size(); i++) {
+        scratch_bvs.push_back(BitVector::create(master.size()));
+        size_t num_this_thread = std::min(num_per_thread, num_iter - num_per_thread*i);
```

Comment:
as far as I can see this will merge too many posting lists if there are more threads than posting lists

Reply:
Fixed